Diagnostic and cmorization for CMUG WP5.4 LST

valeriupredoi opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description

<!--
    Please describe your changes here, especially focusing on why this pull request makes
    ESMValTool better and what problem it solves.


    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/en/latest/community/

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->

This is a PR opened to streamline the process of integration of the CMUG WP5.4-LST data; work is done by @morobking and I have a mere technical support role here.

Also, @hb326 I knows you involved in the CMUG project, maybe you could have a bit of time help out/take a look/review this when ready? :beer: 

- Closes https://github.com/ESMValGroup/ESMValTool/issues/1945
- Link to documentation:

* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [ ] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [ ] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [ ] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [ ] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [ ] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [ ] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [ ] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [ ] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

### [New or updated data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html)

- [ ] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation) is available
- [ ] [🛠][1] The dataset has been [added to the CMOR check recipe](https://docs.esmvaltool.org/en/latest/community/dataset.html#testing)
- [ ] [🧪][2] Numbers and units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#scientific-sanity-check)

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

axel-lauer commented:
I thought it might be worth to copy some of my comments on https://github.com/ESMValGroup/ESMValTool/issues/1945 here, so we can hopefully revive working on this pull request.

I checked out ESMValCore branch https://github.com/ESMValGroup/ESMValCore/tree/ESACCI_LST_uncerts and
ESMValTool branch https://github.com/ESMValGroup/ESMValTool/tree/CMUG-WP5.4-LST
successfully ran the cmorizer: cmorize_obs -c ./config-user.yml -o ESACCI_LST_UNCERT

There are, however, lots of "debug" messages such as "BOOO", "HELLOE!" or warning messages such as "Multiple coordinates for axis Y.". This might require another look.

I also ran recipe_esacci_lst_uncerts.yml : esmvaltool run ./ESMValTool/esmvaltool/recipes/recipe_esacci_lst_uncerts.yml --config_file ./config-user.yml

The ESMValTool preprocessor finishes successfully, but then the script lst_uncert/lst_uncert.py crashes with division by zero.

valeriupredoi commented:
cheers @axel-lauer :beer:

morobking commented:
They spook me into fixing things :)

Has something changed in how v1 to v2 uses the user config yml?
I'm getting this message when running the recipe as it stands:
esmvalcore.experimental.config._validated_config.InvalidConfigParameter: `write_plots` is not a valid config parameter.

morobking commented:
@valeriupredoi 
I've got it so ESMValTool at least tries to run the recipe but it fails with something I'm finding cryptic.
I've pushed the current version of the recipe file to github so you can see; several bits are commented out as I'm still trying to work out how to fix this.
Here is the error I'm getting:
yaml.parser.ParserError: while parsing a block mapping
  in "/home/users/robking/ESMValTool/esmvaltool/recipes/recipe_esacci_lst_uncerts.yml"
, line 57, column 6
did not find expected key
  in "/home/users/robking/ESMValTool/esmvaltool/recipes/recipe_esacci_lst_uncerts.yml"
, line 80, column 7

valeriupredoi commented:
this one's an easy one, Rob! Your variable needs to be indednted:
```yaml
      ts:
          preprocessor: lst_preprocessor
...
```
line 80 is where `ts:` is listed so it falls under `variables` so it and its keys need indentation

morobking commented:
Thanks @valeriupredoi  that's a rookie error made by a rookie :)

Alas that leads me onto the next error, this one has only started since using the latest version of ESMValTool.
When I try to run the recipe I now get this:
KeyError: 'mip'
I can only see two instances of CMIP6 in the yml file, so not sure what this is referring to.

If I can just get to the stage where the python script for the diagnostic runs, I feel like it wont be long until I get this producing some results!

morobking commented:
@valeriupredoi Hi Valeriu, any thoughts on what could be cause the issue in my last comment? Its still occuring, is this another issue around how core and tool are interacting to get the new cmor tables I've needed?

valeriupredoi commented:
Hi @morobking I have a tweaked a little bit your recipe: I added a title (this is mandatory since v2.4), a mip to your variable (although it may not be correct, please check) - you need to specify a mip to variables since some variables can be found in different mips (eg Amon, Omon, Eday etc), and replaced `mip: CMIP6` with `project: CMIP6`, since "CMIP6" is not a mip, but a project, although the whole dataset:
```
        additional_datasets:
          - {dataset: esacci_lst_uncerts, project: CMIP6, type: sat, tier: 2, start_year: 2010, end_year: 2010, version: '1.00'}
```

looks a bit fishy - that should be an OBS dataset so it needs to have `project: OBS` instead. I recommend you read the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/overview.html) on recipe structure and the bits and bobs that are needed to have it correctly configured :+1:

valeriupredoi commented:
also (rather important) heads up, note that the cmorizer interface has changed, have a look at the [new documentation](https://docs.esmvaltool.org/en/latest/develop/dataset.html#create-a-cmorizer-for-the-dataset) and/or ask @remi-kazeroni (our resident OBS cmorizers expert) - the scripts now live in `esmvaltool/cmorizers/data/formatters/datasets/` with slight variations for the import of utilities etc, so you'll have to restructure the addition of the lst cmorizers in line with the new structure (eg for utilities import use `from ...utilities import fix_coords, save_variable` etc) :beer: Let us know if you need any help!

remi-kazeroni commented:
I have also started a Github discussion on the new cmorizer interface with some info on how to adapt cmorizer sitting in PRs to the new interface: https://github.com/ESMValGroup/ESMValTool/discussions/2560 That will be needed before merging this Pr into the main one.

Surely known to you but there is already a [cmorizer](https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/cmorizers/data/formatters/datasets/esacci_lst.py) for ESACCI-LST in the main branch, though with much less variables [available](https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/cmorizers/data/cmor_config/ESACCI-LST.yml).

morobking commented:
@valeriupredoi Thank you for looking at the recipe.
I get this error when trying to run the recipe:
ERROR   Unable to load CMOR table (project) 'CMIP6' for variable 'tsLocalSfcErrDay' with mip 'custom'
This looks like it is referring to this statement:
datasets:
  - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r1i1p1f2, start_year: 2010, end_year: 2010, grid: gn, frequency: monthly}

But that model data won't ever have the tsLocalSfcErrDay variable in it (or any of the other 'new' variables I've defined in the custom CMOR tables).

Ultimately I am going to need the normal ts 'surface temperature' form the models and all of the new CCI LST data variables passed into the diagnostic python file so I can make the comparisons desired.

Am I right in where this is currently failing?
Or is there something else?

The OBS data are in this folder structure:
......./OBS/Tier2/ESACCI_LST_UNCERT/OBS_ESACCI_LST_UNCERTS_sat_1.00_Amon_*.nc
So in this part of the recipe
    variables:

      tsLocalSfcErrDay:
        mip: custom
        preprocessor: lst_preprocessor
        additional_datasets:
          - {dataset: esacci_lst_uncerts, project: OBS, type: sat, tier: 2, start_year: 2010, end_year: 2010, version: '1.00'}
 the project:OBS is right?

morobking commented:
@valeriupredoi I've managed to get the new variables (mostly) to pass through the recipe into the diagnostic code. So I'm almost ready to write the python to make the plots - feels like real progress :) Thanks to @ehogan 
However, there is still one small problem I'm struggling with:
When I try to try to pass tsDay and tsNight from the recipe to the diagnostric, esmvaltool fails with:
esmvalcore.cmor.check.CMORCheckError: There were errors in variable lst:
lst: does not match coordinate rank  in cube:
surface_temperature / (kelvin)      (time: 1; latitude: 3600; longitude: 7200)
     Dimension coordinates:
          time                           x            -                -
          latitude                       -            x                -
          longitude                      -            -                x
     Attributes:
          Conventions: CF-1.7
          source_file: /work/scratch-nopw/morobking/OBS/Tier2/ESACCI_LST_UNCERTS/OBS_ESACCI_L...
          var: tsNight

Can you suggest what might be going on here please? I suspect it is a CMORizer issue? 
In esmvaltool/cmorizers/obs/cmorize_obs_esacci_lst_uncert.py I dont make any explicit reference to lst in making the files.
esmvalcore/cmor/tables/custom/CMOR_tsDay.dat and esmvalcore/cmor/tables/custom/CMOR_tsNight.dat I have updated, so I guess I need to rerun the CMORizer, and now thats going to need the work @remi-kazeroni talks about? 
Because now I get this error when trying to run the CMORizer:
ModuleNotFoundError: No module named 'esmvaltool.cmorizers.obs.cmorize_obs'

Cheers

remi-kazeroni commented:
> Can you suggest what might be going on here please? I suspect it is a CMORizer issue? In esmvaltool/cmorizers/obs/cmorize_obs_esacci_lst_uncert.py I dont make any explicit reference to lst in making the files. esmvalcore/cmor/tables/custom/CMOR_tsDay.dat and esmvalcore/cmor/tables/custom/CMOR_tsNight.dat I have updated, so I guess I need to rerun the CMORizer, and now thats going to need the work @remi-kazeroni talks about? Because now I get this error when trying to run the CMORizer: ModuleNotFoundError: No module named 'esmvaltool.cmorizers.obs.cmorize_obs'

Hi @morobking, if you had cmorized the ESACCI-LST data in the past and do not need to change anything, I think you do not need to worry now about the new cmorizer interface. You would need to update your cmorizer to the new interface now only if you need to rerun it and you have fully switched to ESMValTool version 2.5. To check that the new interface is working for you, you could type `esmvaltool data --help` and see if that works fine. To update your cmorizer, you would need to place the script in `esmvaltool/cmorizers/data/formatters/datasets/` and name it `esacci_lst_uncert.py` (no more cmorize_obs needed). 
Aslo, you need to move `esmvaltool/cmorizers/obs/cmor_config/ESACCI_LST_UNCERT.yml` to `esmvaltool/cmorizers/data/cmor_config/ESACCI_LST_UNCERT.yml`. The main changes to the script are:
- the utilities import syntax from [this line](https://github.com/ESMValGroup/ESMValTool/blob/b35108c05994c748e528c9e2dd4e334d78e63f03/esmvaltool/cmorizers/data/formatters/cmorize_obs_esacci_lst_uncert.py#L23). See a working example [here](https://github.com/ESMValGroup/ESMValTool/blob/CMUG-WP5.4-LST/esmvaltool/cmorizers/data/formatters/datasets/esacci_lst.py#L30).
- the call to the `cmorization function` from [this line](https://github.com/ESMValGroup/ESMValTool/blob/b35108c05994c748e528c9e2dd4e334d78e63f03/esmvaltool/cmorizers/data/formatters/cmorize_obs_esacci_lst_uncert.py#L27). See working example [here](https://github.com/ESMValGroup/ESMValTool/blob/b35108c05994c748e528c9e2dd4e334d78e63f03/esmvaltool/cmorizers/data/formatters/datasets/esacci_lst.py#L35).

The examples I used are directly taken from your previous work on the ESACCI-LST data. We had updated this cmorizer to the new interface. You could directly compare the 2 cmorizers. 

If I understand your work here correctly, you are planning on adding support to a new dataset (named ESACCI-LST-UNCERT) instead of extending the existing one (ESACCI-LST), right? Wouldn't it be simpler to extend the ESACCI-LST cmorizer?

morobking commented:
@remi-kazeroni Turns out I do need to run the CMORizer now to fix something in the data.
I've updated to V2.5.
I'm running the command:
esmvaltool data format -config_file config_morobking.yml ESACCI_LST_UNCERT
but get
ERROR   Could not find formatter for ESACCI_LST_UNCERT

My cmorizer python is at
esmvaltool/cmorizers/data/formatters/datasets/esacci_lst_uncert.py
and the RAWOBS data is at
RAWOBS/Tier2/ESACCI_LST_UNCERT/

Is there something else I need to update or move to get this working?
I know the python code its self wont work at this stage but the errors I'm getting imply to me that esmvaltool isnt getting that far yet.

morobking commented:
ok I've got a successful v2.5 install
and so far almost everything is now working again.

But when running the recipe I get this error
esmvalcore.cmor.check.CMORCheckError: There were errors in variable ts:
ts: does not match coordinate rank
 in cube:
surface_temperature / (kelvin)      (time: 1; latitude: 3600; longitude: 7200)
    Dimension coordinates:
        time                             x            -                -
        latitude                         -            x                -
        longitude                        -            -                x
    Attributes:
        Conventions                 CF-1.7
        source_file                 /work/scratch-nopw/morobking/OBS/Tier2/ESACCI_LST_UNCERTS/OBS_ESACCI_L...
        var                         tsDay

The other files/variables all seem to work ok and I can get them through to the diagnostic python, but tsDay and tsNight are proving elusive.
Any thoughts @remi-kazeroni  @valeriupredoi  please?

morobking commented:
@axel-lauer Thanks for the email Axel.
I have the above question about esmvalcore.cmor.check.CMORCheckError
and
related to this PR,
I want to be able to get multiple model ensemble members into the diagnostic using for example
  ts:
        mip: Amon
        preprocessor: lst_preprocessor
        additional_datasets:
          - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r(1:2)i1p1f2, start_year: 2010, end_year: 2010, grid: gn}
but I only see one instance of ts when I look at the data the diagnostic is loading.
If I use a seperate dataset section in the recipe it looks for all the variables I know arent available in the models that are only in the CCI data. Is there a solution for this?

Finally for now, I'm getting this error in a seperate piece of work
_intersect_modulus
    raise ValueError(
ValueError: coordinate's range greater than coordinate's unit's modulus

This is coming from trying to get ERA5-Land data to load in a diagnostic. I see only the pr variable has a cmor fix, so have written something to quickly cmorizse ts and ts. That all runs fine, including util.fix_coord on time latitude and longitude. But when the recipe runs I  get that Value Error when it tries to extract a region in the preprocessor.

axel-lauer commented:
> I want to be able to get multiple model ensemble members into the diagnostic using for example ts: mip: Amon preprocessor: lst_preprocessor additional_datasets: - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r(1:2)i1p1f2, start_year: 2010, end_year: 2010, grid: gn} but I only see one instance of ts when I look at the data the diagnostic is loading. If I use a seperate dataset section in the recipe it looks for all the variables I know arent available in the models that are only in the CCI data. Is there a solution for this?

After updating https://github.com/ESMValGroup/ESMValCore/tree/ESACCI_LST_uncerts with the latest `main` I could successfully run both of these two options:

```
      ts:
        mip: Amon
        preprocessor: lst_preprocessor
        additional_datasets:
          - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r1i1p1f2, start_year: 2010, end_year: 2010, grid: gn}
          - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r2i1p1f2, start_year: 2010, end_year: 2010, grid: gn}
```

and

```
      ts:
        mip: Amon
        preprocessor: lst_preprocessor
        additional_datasets:
          - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r(1:2)i1p1f2, start_year: 2010, end_year: 2010, grid: gn}
```

Both versions run successfully and produce the same plot:

![timeseries](https://user-images.githubusercontent.com/25367775/163961881-dbbc32bf-3a48-42a0-bca6-4d82a91a0614.png)

If I am not mistaken, I would think from the log file [log.txt] (https://github.com/ESMValGroup/ESMValTool/files/8511364/log.txt) that both ensemble members are read by the diagnostic. So maybe simply updating ESMValCore with the latest `main` branch does the trick? It could also be that I do not fully understand the actual problem... In that case (which is totally possible) please provide some more instructions on how to reproduce the problem.

> Finally for now, I'm getting this error in a seperate piece of work _intersect_modulus raise ValueError( ValueError: coordinate's range greater than coordinate's unit's modulus Is this related to [SciTools/iris#3907](https://github.com/SciTools/iris/issues/3907) where I see ESMValTool comes up that conversation?
> 
> This is coming from trying to get ERA5-Land data to load in a diagnostic. I see only the pr variable has a cmor fix, so have written something to quickly cmorizse ts and ts. That all runs fine, including util.fix_coord on time latitude and longitude. But when the recipe runs I get that Value Error when it tries to extract a region in the preprocessor.

@morobking Could you guide me to a recipe I could use for testing? I will have a look then.

morobking commented:
Thank you @axel-lauer 

On the multiple ensemble question:
I have updated the branch with git pull
In the diagnostic python esmvaltool/diag_scripts/lst_uncert/lst_uncert.py 
I have these lines to load the data
    input_metadata = config['input_data'].values()

    loaded_data = {}
    ancestor_list = []
    for dataset, metadata in group_metadata(input_metadata, 'dataset').items():
        cubes, ancestors = _get_input_cubes(metadata)
        loaded_data[dataset] = cubes

When I do a print(loaded_data) I get
{'UKESM1-0-LL': {'ts': <iris 'Cube' of surface_temperature / (K) (time: 12; latitude: 8; longitude: 5)>}, 
'ESACCI_LST_UNCERTS': {'tsLocalSfcErrDay': <iris 'Cube' of uncertainty from locally correlated errors on surface scales daytime / (K) (time: 12; latitude: 200; longitude: 200)>, 'tsLocalSfcErrNight': <iris 'Cube' of uncertainty from locally correlated errors on surface scales night time / (K) (time: 12; latitude: 200; longitude: 200)>, 'tsLocalAtmErrDay': <iris 'Cube' of uncertainty from locally correlated errors on atmospheric scales / (K) (time: 12; latitude: 200; longitude: 200)>, 'tsLocalAtmErrNight': <iris 'Cube' of uncertainty from locally correlated errors on atmospheric scales / (K) (time: 12; latitude: 200; longitude: 200)>, 'tsLSSysErrDay': <iris 'Cube' of uncertainty from large-scale systematic errors / (K) (time: 12)>, 'tsLSSysErrNight': <iris 'Cube' of uncertainty from large-scale systematic errors / (K) (time: 12)>, 'tsUnCorErrDay': <iris 'Cube' of uncertainty from uncorrelated errors / (K) (time: 12; latitude: 200; longitude: 200)>, 'tsUnCorErrNight': <iris 'Cube' of uncertainty from uncorrelated errors / (K) (time: 12; latitude: 200; longitude: 200)>}}

Here there is only one entry for UKESM data but in the recipe
  ts:
        mip: Amon
        preprocessor: lst_preprocessor
        additional_datasets:
          - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r1i1p1f2, start_year: 2010, end_year: 2010, grid: gn}
          - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r2i1p1f2, start_year: 2010, end_year: 2010, grid: gn}

The same output occurs for print(loaded_data) when I use the r(1:2)i1p1f2 syntax or other CMIP models in the recipe when defining the ts variable.

My question is how to get the different ensemble members available in the python code of the diagnostic?

I'll keep the other 2 questions seperate.

axel-lauer commented:
> On the multiple ensemble question: I have updated the branch with git pull In the diagnostic python esmvaltool/diag_scripts/lst_uncert/lst_uncert.py I have these lines to load the data input_metadata = config['input_data'].values()
> 
> ```
> loaded_data = {}
> ancestor_list = []
> for dataset, metadata in group_metadata(input_metadata, 'dataset').items():
>     cubes, ancestors = _get_input_cubes(metadata)
>     loaded_data[dataset] = cubes
> ```
> Here there is only one entry for UKESM data but in the recipe ts: mip: Amon preprocessor: lst_preprocessor additional_datasets: - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r1i1p1f2, start_year: 2010, end_year: 2010, grid: gn} - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical, ensemble: r2i1p1f2, start_year: 2010, end_year: 2010, grid: gn}
> 
> The same output occurs for print(loaded_data) when I use the r(1:2)i1p1f2 syntax or other CMIP models in the recipe when defining the ts variable.
> 
> My question is how to get the different ensemble members available in the python code of the diagnostic?

Thanks for the additional explanation, I think now I get it. I believe the problem is in function `_get_input_cubes` in `lst_uncert.py` and not in the recipe or preprocessor. As the dictionary keys used in `_get_input_cubes` for each dataset are simply the variable name (`short_name`), entries for different ensemble members are overwritten for UKESM1-0-LL. I am not exactly sure what is supposed to happen later in the code, but an option might be to use a list of cubes (instead of a dictionary) and add new items with `append()` or altenatively, use a more unique key for the dictionary than just `short_name`, possibly containing the ensemble member if present.

morobking commented:
@axel-lauer Thank you, that solves the first problem :+1: 

My second question is around the new tsDay and tsNight vatiables I have had to make from the CCI LST data.
When I run the recipe with them included I get this error:
  File "/home/users/robking/ESMValCore/esmvalcore/cmor/check.py", line 235, in report_errors
    raise CMORCheckError(msg)
esmvalcore.cmor.check.CMORCheckError: There were errors in variable ts:
ts: does not match coordinate rank
 in cube:
surface_temperature / (kelvin)      (time: 1; latitude: 3600; longitude: 7200)
    Dimension coordinates:
        time                             x            -                -
        latitude                         -            x                -
        longitude                        -            -                x
    Attributes:
        Conventions                 CF-1.7
        source_file                 /work/scratch-nopw/morobking/OBS/Tier2/ESACCI_LST_UNCERTS/OBS_ESACCI_L...
        var                         tsDay


I do not get this for the other new CCI LST uncertainty variables, like tsLSErrDay etc...
Is there something to be looking for to fix, or something I can do in the recipe?

The third question concerns a different branch, so I'll start a conversation there about that soon.

axel-lauer commented:
> My second question is around the new tsDay and tsNight vatiables I have had to make from the CCI LST data. When I run the recipe with them included I get this error: File "/home/users/robking/ESMValCore/esmvalcore/cmor/check.py", line 235, in report_errors raise CMORCheckError(msg) esmvalcore.cmor.check.CMORCheckError: There were errors in variable ts: ts: does not match coordinate rank in cube: surface_temperature / (kelvin) (time: 1; latitude: 3600; longitude: 7200) Dimension coordinates: time x - - latitude - x - longitude - - x Attributes: Conventions CF-1.7 source_file /work/scratch-nopw/morobking/OBS/Tier2/ESACCI_LST_UNCERTS/OBS_ESACCI_L... var tsDay
> 
> I do not get this for the other new CCI LST uncertainty variables, like tsLSErrDay etc... Is there something to be looking for to fix, or something I can do in the recipe?

I think there are some possible reasons for this problem. Here is what I would do:

ESMValTool:
1) Change the variable name in the cmorized files from "lst" to "tsDay" / "tsNight" and rerun the cmorizer.

ESMValCore:

2) Make sure that the dimensions in the custom cmor tables CMOR_tsDay.dat and CMOR_tsNight.dat are correct. At the moment, the tables say "dimensions: time". I believe this should rather be something like "dimensions: longitude latitude time". Btw. maybe also check if the ordering of the dimensions in your other custom CMOR_tables (e.g. CMOR_tsLocalSfcErr.dat") is correct.

I think the most important thing to check is 2), item 1) is probably rather a cosmetic problem.

morobking commented:
@axel-lauer Thank you - your point 2 is the answer! Not sure I would have found that, nor sure how I missed that when originally setting up the cmor tables.

My third question is about a something on a different branch, I'll set up a draft PR there.

MargaridaS commented:
Hello @axel-lauer

I contact you because I found another dataset with the same error `ValueError: coordinate's range greater than coordinate's unit's modulus` than the one obtained by @morobking:
> 
> Finally for now, I'm getting this error in a seperate piece of work _intersect_modulus raise ValueError( ValueError: coordinate's range greater than coordinate's unit's modulus Is this related to [SciTools/iris#3907](https://github.com/SciTools/iris/issues/3907) where I see ESMValTool comes up that conversation?
> 
> This is coming from trying to get ERA5-Land data to load in a diagnostic. I see only the pr variable has a cmor fix, so have written something to quickly cmorizse ts and ts. That all runs fine, including util.fix_coord on time latitude and longitude. But when the recipe runs I get that Value Error when it tries to extract a region in the preprocessor.

The dataset for my case is Ghrsstostia-v2.0, launched with the recipe: 
[OBS_ostia_2007_extract_region.txt](https://github.com/ESMValGroup/ESMValTool/files/9092987/OBS_ostia_2007_extract_region.txt) and with the file: [tos_200701.zip](https://github.com/ESMValGroup/ESMValTool/files/9092995/tos_200701.zip). Is this a bug? How could it be solved?

Thank you very much in advance.

Kind regards,

Marga

MargaridaS commented:
> @MargaridaS I am sorry, but I cannot remember the details. I believe it was @valeriupredoi and @zklaus helping. Can one of you remember the problem with the coordinates? I briefly checked your file tos_200701.nc and I saw the longitudes are running form -180 to 180. I believe CMOR standard would be 0 to 360 instead. Not sure if that causes the problem but it might be worth checking.

Hi @axel-lauer 

Thank you for your prompt answer. I changed longitudes to 0 to 360 but the same error appears. I see a bug was reported to iris some time ago: https://github.com/SciTools/iris/issues/3907 . Could this be a type of variables error?

Best regards,

Marga

MargaridaS commented:
Hi @axel-lauer 

Setting manually the first bound to zero and the last bound to 360 solved the issue. 
Thank you very much for your attention.

Best,
Marga
